test_common.py

```diff
@@ -0,0 +1,30 @@
+import pathlib
+import pytest
+import os
+
+import common
+
+
+def test_atomic_write(tmp_path: pathlib.Path) -> None:
+
+    user = os.geteuid()
+    group = os.getegid()
+
+    fname = tmp_path / "file1"
```

Comment:
It is not clear what is being tested in each case. Maybe a small commentary header would help the reader.

Reply:
Even better: use functions with proper names so that our view when reading the function isn't cluttered by `with open...`.

Reply:
The test function is called `test_atomic_write()` in `test_common.py`.  It tests `common.atomic_write()`.

Optimally, the code should should be clear without a comment. I refactored the test a bit. Is that better?

Reply:
Is this test just doing normal file writes and reads? It isn't really testing the atomic nature of the call (i think). This is what puzzles me.

Reply:
hm. `common.atomic_write()` is a context manager that returns a file object. you can use it like
```python
     with common.atomic_write(filename) as file:
          file.write(content)
```
similar to 
```python
     with open(filename, "w") as file:
          file.write(content)
```

the difference is, that it first creates a temporary file (`tempfile.mkstemp()`). Then it returns that file. Afterwards, it will (atomically) rename the file to the requested filename. The point is not to modify the existing file, but replace it as a whole. Other processes won't see partial writes.

The test just does that 3 times, with different parameters.

Could I do something to make that clearer?

Reply:
(maybe the name is bad... it's taken from similar code on the internet).

Reply:
> It isn't really testing the atomic nature of the call 

True, it doesn't fully test that an intermediate file was created, and that the file was never seen partially written. It merely tests, that afterwards, the expected content is in the file.

Reply:
what about https://pypi.org/project/atomicwrites/ ?

Reply:
yes, that serves the same purpose.

Note the [comment](https://github.com/untitaker/python-atomicwrites?tab=readme-ov-file#unmaintained) about being unmaintained. That might however not matter much. What is there to "maintain"?

To me, the functionality is complex enough that I would add a separate function and unit tests. On the other hand, it is *not* complex enough, to warrant dragging in another dependency (I'd just re-implement it). Also the re-implementation is simpler, because it only targets Python3.11+ and Linux.

Before using another module, I would rather fallback to plain `open(filename, "w")`. I admit, even adding a `common.atomic_write()` is borderline too much for CDA.

I would not use this module. But sure, that would also work. I guess, I often prefer implementing things ourselves over using external code from unknown people. But that is only a personal preference.

Reply:
I think a file like /etc/resolv.conf, benefits from being written atomically so that clients don't see intermediate states. Otherwise, resolving names might fail at a particular point in time when CDA is running. It's very (extremely?) unlikely, but if you encounter a problem with it, it's also almost impossible to debug/reproduce.

Granted, somebody can be of the opinion that this is no problem, and better to have less code in CDA.

Coincidentally, as this helper seems to stir more controversy/discussion/questions, the burden increases. I didn't anticipate this to be a controversial thing...

IMO it's fine (preferable over not having it). Opinions may differ. Should I drop it?

Reply:
@SalDaniele Up to you.

Reply:
I think it should be ok if you comment on the purpose of it in the atomic code such that noone is left guessing: "why is this needed". The tests are good, but only if you note through commentary that you aren't checking for the intermediate file. As a reader of the code and without much context I just found it a little difficult.

Reply:
The test should be:
```
     with atomic.open(filename, "w") as file:
          file.write(content)
          file.flush() # or something similar
          with open(filename, 'r') as file:
               # assert that you can't read it
```

That's how you would test an atomic writer. I'll still leave the decision to keep or remove up to @SalDaniele .

Reply:
Do we care that we have an atomic writer/reader or do we care about the behavior that multiple concurrent updates to DNS end up being serialized? iow, are you testing the right thing?

Reply:
> The test should be:

Unit test is extended.